### ISSUE TITLE
Some particles use itemstack as particle data

### ISSUE THREAD

**Issue by TheComputerGeek2** [Slikey/EffectLib#34]
The particle effect, icon crack, uses an ItemStack as its data rather than just a material type, I noticed that effectlib has not yet updated to accommodate this change.

What can I do to help address the issue?

**Comment by NathanWolf** [comment]
I'm not certain what you mean here. EffectLib doesn't use Spigot's particle API, it directly sends packets. Are you saying the packet format has changed? (That would not make much sense to me, I can't really believe the particle packet can now contain an entire item stack...)

Is there something not working that was working before? I don't use the icon crack effect much myself, so it may be broken and I haven't noticed.

**Comment by TheComputerGeek2** [comment]
@NathanWolf well, I noticed that in spigot, the icon crack effect now wants to be passed an itemstack, and when effectlib sends an itemcrack effect, anyone who sees it gets immediately disconnected with a packet error. So it seems that the packet format has changed.

**Comment by NathanWolf** [comment]
It doesn't appear to have changed, unless the protocol wiki hasn't been updated:

https://github.com/Slikey/EffectLib/issues/34#issuecomment-254393392

Are you certain you're not just sending an invalid material/data combination? Was what you're doing working in a previous server version?

**Comment by NathanWolf** [comment]
I just dug into Spigot's code, I don't think anything has changed here other than they added an API.

```
if (particle.getDataType().equals(ItemStack.class)) {
            if (obj == null) {
                return new int[]{0, 0};
            }
            ItemStack itemStack = (ItemStack) obj;
            return new int[]{itemStack.getType().getId(), itemStack.getDurability()};
        }
```

Spigot just takes the material id and durability out of the ItemStack and puts it in the particle packet as an array, same as EffectLib is doing.

Have you narrowed this down to some specific code, effect, or combination of material/data that is crashing?

**Comment by TheComputerGeek2** [comment]
@NathanWolf the crash is client side, and I have not been able to narrow it down to a specific combination of material/data, however, I can provide numerous examples that have caused a client disconnect from packet error. I have been using effectlib's built in config handler for generating the effects. If it helps, this issue has been around for a bit. Do you want me to try getting the client disconnect screen?

**Comment by NathanWolf** [comment]
If it shows the particle in question (and especially if it shows the data), then yes.

Do you have examples of configurations you're using that would include the icon_crack particle? Are you generating the data for this particle dynamically, like from an in-game item? If not, then it should be possible to see what you might be sending, right?

**Comment by TheComputerGeek2** [comment]
The data is not being dynamically generated, here is an example of one of the data sets that caused a client disconnect.

```yaml
class: AnimatedBallEffect
particle: ITEM_CRACK
speed: 0.02
xFactor: 1
yFactor: 2
zFactor: 1
size: 1
iterations: -1
material: COOKIE
materialData: 0
period: 1
particlesPerIteration: 5
particles: 150
```
At one point, this was working, with the same set of parameters. I can't remember which minecraft update this broke on.
I'll work on getting the disconnect message now.

**Comment by LCastr0** [comment]
Can you please test the same configuration with a different ItemStack, that stack being WOOL? And one with the stack being STICK? I might be completely wrong but I don't think icon crack supports iteme anymore. Only blocks. As I said, I might be wrong, but doesn't hurt to test

**Comment by NathanWolf** [comment]
I will try it when I can.

That said, I was pretty sure this particle is only supposed to be used for tools. It's for when the tool breaks. So it doesn't necessarily surprise me that it can't be used with any arbitrary item.

Can you try it with a diamond_axe or similar?

**Comment by TheComputerGeek2** [comment]
@LCastr0 I know for a fact that it used to support items as this effect used to work fine, I can't see why it would stop supporting them, and I know that it was also successful in the past when using golden apples for example.

@NathanWolf I'll test that out now, just give me a few minutes and I'll check in with the results.

**Comment by NathanWolf** [comment]
This seems to be an internal Spigot/mcserver issue. I can reproduce using the Spigot API, simple as this:
```
event.getPlayer().playEffect(event.getPlayer().getEyeLocation(), Effect.ITEM_BREAK, Material.DIAMOND_AXE);
```

I'm wondering if this particle does not expect to have a data byte attached, I'll play around a little and see if I can fix this in EffectLib.

**Comment by NathanWolf** [comment]
I have fixed this in EffectLib 5.1, which I will publish to DBO shortly, or you can grab it from my CI:

http://jenkins.elmakers.com/job/EffectLib/91/de.slikey$EffectLib/
